File: 2014_May/Zig/cimek.zig
const std = @import("std");
const print = std.debug.print;
const workingDir = std.fs.cwd();

pub fn dynamicSearch(elements: *[][]const u8, index: usize, direction: u8) u8 {
    if (index < 0 and index > 7) {
        return 0;
    } else if (std.mem.eql(u8, elements.*[index], "0") and direction == '0') {
        return 1 + dynamicSearch(elements, index - 1, '-') + dynamicSearch(elements, index + 1, '+');
    } else if (std.mem.eql(u8, elements.*[index], "0") and direction == '-') {
        return 1 + dynamicSearch(elements, index - 1, '-');
    } else if (std.mem.eql(u8, elements.*[index], "0") and direction == '+') {
        return 1 + dynamicSearch(elements, index + 1, '+');
    } else {
        return 0;
    }
}

pub fn forceAbbreviate(allocator: *const std.mem.Allocator, ip: *const []const u8) ![]const u8 {
    var splittedAddress = std.mem.splitSequence(u8, ip.*, ":");
    var buffer = std.ArrayList([]const u8).init(allocator.*);

    while (splittedAddress.next()) |addressPart| {
        try buffer.append(addressPart);
    }

    var i: usize = 0;
    var longest: u8 = 0;

    for (buffer.items, 0..) |_, index| {
        const amount = dynamicSearch(&buffer.items, index, '0');

        if (amount > longest and amount > 1) {
            longest = amount;
            i = index;
        }
    }

    for (0..longest) |index| {
        if (index == longest - 1) {
            buffer.items[i] = "";
        } else {
            _ = buffer.orderedRemove(i);
        }
    }

    return try std.mem.join(allocator.*, ":", try buffer.toOwnedSlice());
}

pub fn removeInitialZeros(allocator: *const std.mem.Allocator, ip: *const []const u8) ![]const u8 {
    var requiredParts = std.ArrayList([]const u8).init(allocator.*);
    var splittedAddress = std.mem.splitSequence(u8, ip.*, ":");

    while (splittedAddress.next()) |addressPart| {
        var terminateCondition = false;

        for (addressPart, 1..) |alphanumeric, count| {
            if (terminateCondition == false and alphanumeric == '0' and count < 4) {
                continue;
            } else {
                terminateCondition = true;
                var currentValue = try allocator.dupe(u8, &[_]u8{ alphanumeric });
                try requiredParts.append(currentValue);
            }
        }

        var delimiter = try allocator.dupe(u8, &[_]u8{ ':' });
        try requiredParts.append(delimiter);
    }

    requiredParts.shrinkAndFree(requiredParts.items.len - 1);
    return try std.mem.concat(allocator.*, u8, try requiredParts.toOwnedSlice());
}

pub fn moreThanSeventeenZeros(ip: *const []const u8) bool {
    var counter: u32 = 0;

    for (ip.*) |alphanumeric| {
        if (alphanumeric == '0') {
            counter += 1;
        }
    }

    return if (counter >= 18) true else false;
}

pub fn main() !void {
//
// Task : 1
//
    const inputFile = try workingDir.openFile("ip.txt", .{});
    defer inputFile.close();
    
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator(); 
    
    var ipAddresses = std.ArrayList([]const u8).init(allocator); 
    var bufferedReader = std.io.bufferedReaderSize(41, inputFile.reader());
    var reader = bufferedReader.reader();

    while (try reader.readUntilDelimiterOrEof(&bufferedReader.buf, '\n')) |line| {
        const trimmedLine = std.mem.trimRight(u8, line, "\r");
        var heapBuffer = try allocator.dupe(u8, trimmedLine);
        try ipAddresses.append(heapBuffer);
    }
//
// Task : 2
//
    print("2. feladat:\n", .{});
    print("Az allomanyban {d} darab adatsor van.\n\n", .{ipAddresses.items.len});
//
// Task : 3
//
    var lowestAddress: []const u8 = undefined;
    var best: u32 = std.math.maxInt(u32);

    for (ipAddresses.items) |item| {
        var counter: u32 = 0;

        for (item) |alphanumeric| {
            counter += alphanumeric;
        }

        if (counter < best) {
            best = counter;
            lowestAddress = item;
        }
    }

    print("3. feladat:\n", .{});
    print("A legalacsonyabb tarolt IP-cim:\n", .{});
    print("{s}\n\n", .{lowestAddress});
//
// Task : 4
//
    var ipTypesAmount = [_]u32{ 0, 0, 0 };

    for (ipAddresses.items) |item| {
        if (std.mem.eql(u8, item[0..9], "2001:0db8")) {
            ipTypesAmount[0] += 1;
        } else if (std.mem.eql(u8, item[0..7], "2001:0e")) {
            ipTypesAmount[1] += 1;
        } else {
            ipTypesAmount[2] += 1;
        }
    }    

    print("4. feladat:\n", .{});
    print("Dokumentacios cim: {d} darab\n", .{ipTypesAmount[0]});
    print("Globalis egyedi cim: {d} darab\n", .{ipTypesAmount[1]});
    print("Helyi egyedi cim: {d} darab\n\n", .{ipTypesAmount[2]});
//
// Task : 5
//
    const outputFile = try workingDir.createFile("sok.txt", .{});
    defer outputFile.close();

    var writer = outputFile.writer();

    for (ipAddresses.items, 1..) |item, index| {
        if (moreThanSeventeenZeros(&item) == true) {
            try std.fmt.format(writer, "{d} {s}\n", .{index, item});
        }
    }
//
// Task : 6
//
    print("6. feladat:\n", .{});
    print("Kerek egy sorszamot: ", .{});
    
    const stdin = std.io.getStdIn().reader();
    const userInput = try stdin.readUntilDelimiterOrEof(&bufferedReader.buf, '\n');
    var chosenAddress: []const u8 = undefined;

    if (userInput) |index| {
        chosenAddress = ipAddresses.items[(std.fmt.parseUnsigned(u32, index, 10) catch 1) - 1];
    } else {
        chosenAddress = ipAddresses.items[0];
    }

    const abbreviatedAddress = try removeInitialZeros(&allocator, &chosenAddress);

    print("{s}\n", .{chosenAddress});
    print("{s}\n\n", .{abbreviatedAddress});
//
// Task : 7
//
    const evenMoreAbbreviatedAddress = try forceAbbreviate(&allocator, &abbreviatedAddress);
    print("7. feladat:\n", .{});

    if(std.mem.eql(u8, abbreviatedAddress, evenMoreAbbreviatedAddress)) {
        print("Nem roviditheto tovabb.\n", .{});
    } else {
        print("{s}\n", .{evenMoreAbbreviatedAddress});
    }  
}

File: 2014_May/C#/cimek.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using static System.Console;

class Cimek
{
	public static int DynamicSearch(int index, ref string[] segments, char direction)
	{
		if (index < 0 || index > 7)
		{
			return 0;
		}
		else if (segments[index] == "0" && direction == '0')
		{
			return 1 + DynamicSearch(index - 1, ref segments, '-') + DynamicSearch(index + 1, ref segments, '+');
		}
		else if (segments[index] == "0" && direction == '-')
		{
			return 1 + DynamicSearch(index - 1, ref segments, '-');
		}
		else if (segments[index] == "0" && direction == '+')
		{
			return 1 + DynamicSearch(index + 1, ref segments, '+');
		}
		else
		{
			return 0;
		}
	}

	public static string HardAbbreviate(string ip)
	{
		var addressParts = ip.Split(':');
		var best = 0;
		var index = -1;

		for (int i = 0; i < addressParts.Length; ++i)
		{
			var counter = DynamicSearch(i, ref addressParts, '0');

			if (counter > 1 && counter > best)
			{
				best = counter;
				index = i;
			}
		}

		var result = new List<string>();
		var skipElements = false;

		for (int i = 0; i < addressParts.Length; ++i)
		{
			if (i == index)
			{
				skipElements = true;
			}

			if (!skipElements)
			{
				result.Add(addressParts[i]);
			}
			else
			{
				--best;

				if (best == 0)
				{
					result.Add("");
					skipElements = false;
				}
			}
		}

		return String.Join(":", result.ToArray());
	}

	public static string SoftAbbreviate(string ip)
	{
		var builder = new StringBuilder();

		foreach (var addressPart in ip.Split(':'))
		{
			if (addressPart.StartsWith("000"))
			{
				builder.Append(addressPart[3]);
			}
			else if (addressPart.StartsWith("00"))
			{
				builder.Append(addressPart.AsSpan().Slice(2));
			}
			else if (addressPart.StartsWith("0"))
			{
				builder.Append(addressPart.AsSpan().Slice(1));
			}
			else
			{
				builder.Append(addressPart);
			}

			builder.Append(':');
		}

		builder.Remove(builder.Length - 1, 1);
		return builder.ToString();
	}

	static void Main(string[] args)
	{
	//
	// Task : 1
	//
		var ipAddresses = new List<string>();
		using var reader = new StreamReader("ip.txt");

		while (reader.EndOfStream == false)
		{
			var line = reader.ReadLine() ?? String.Empty;

			if (line.Equals(String.Empty) == false)
			{
				ipAddresses.Add(line.TrimEnd('\r'));
			}
		}
	//
	// Task : 2
	//
		WriteLine("2. feladat:");
		WriteLine($"Az allomanyban {ipAddresses.Count} darab adatsor van.\n");
	//
	// Task : 3
	//
		var lowest = ipAddresses.Select(ip => new
		{
			Address = ip,
			AsciiValue = ip.Sum(character => character),
		})
		.MinBy(ip => ip.AsciiValue);
		
		WriteLine("3. feladat:");
		WriteLine("A legalacsonyabb tarolt IP-cim:");
		WriteLine($"{lowest?.Address}\n");
	//
	// Task : 4
	//
		var documentAddress = ipAddresses.Where(ip => ip.StartsWith("2001:0db8")).Count();
		var globalAddress = ipAddresses.Where(ip => ip.StartsWith("2001:0e")).Count();
		var localAddress = ipAddresses.Where(ip => ip.StartsWith("fc") || ip.StartsWith("fd")).Count();
		
		WriteLine("4. feladat:");
		WriteLine($"Dokumentacios cim: {documentAddress} darab");
		WriteLine($"Globalis egyedi cim: {globalAddress} darab");
		WriteLine($"Helyi egyedi cim: {localAddress} darab\n");
	//
	// Task : 5
	//
		using var writer = new StreamWriter("sok.txt");

		for (int i = 0; i < ipAddresses.Count; ++i)
		{
			if (ipAddresses[i].Count(character => character == '0') >= 18)
			{
				writer.WriteLine($"{i + 1} {ipAddresses[i]}");
			}
		}
	//
	// Task : 6
	//
		WriteLine("6. feladat:");
		Write("Kerek egy sorszamot: ");

		var choice = -1;
		var userInput = ReadLine() ?? String.Empty;
		
		if (Int32.TryParse(userInput, out choice) == false || choice < 1 || choice > ipAddresses.Count)
		{
			choice = 1;
		}

		var abbreviatedAddress = SoftAbbreviate(ipAddresses[choice - 1]);

		WriteLine($"{ipAddresses[choice - 1]}");
		WriteLine($"{abbreviatedAddress}\n");
	//
	// Task : 7
	//
		WriteLine("7. feladat:");

		var evenMoreAbbreviatedAddress = HardAbbreviate(abbreviatedAddress);

		WriteLine("{0}", evenMoreAbbreviatedAddress.Equals(abbreviatedAddress) ? "Nem roviditheto tovabb." : evenMoreAbbreviatedAddress);
	}
}
